### PR TITLE
Optimize `jstest` runner

### LINKED ISSUE #322: Optimize the `jstest` runner
Currently we spin up a container for each file we run, which doesn't scale. We could, in theory build an `eval` statement and load each file which would only require a single container. This assumes tests are self contained and do not depend on one another.

### PR DESCRIPTION
**Pull request by b1ron**
Closes #322.

### CODE REVIEW

**Comment by b1ron** [comment]
The linter is complaining about the type assertion on the error.

File: internal/jstest/jstest.go [hunk 1]
```diff
@@ -53,27 +53,44 @@ func Run(ctx context.Context, dir string, args []string) (*internal.TestResults,
...
-				return nil, err
+	type item struct {
+		file string
+		err  error
+		out  []byte
+	}
+
+	var volume = "tests"
+
+	ch := make(chan item, len(files))
```
Comment (AlekSi) [inline internal/jstest/jstest.go:64]:
```suggestion
	ch := make(chan *item, len(files))
```

To pass only a pointer, not the whole value

File: internal/jstest/jstest.go [hunk 2]
```diff
@@ -53,27 +53,44 @@ func Run(ctx context.Context, dir string, args []string) (*internal.TestResults,
...
+
+	for _, f := range files {
+		go func(f string) {
+			var it item
+			it.file = f
```
Comment (AlekSi) [inline internal/jstest/jstest.go:69]:
```suggestion
			it := &item{
			  file: f,
			}
```

File: internal/jstest/jstest.go [hunk 3]
```diff
@@ -53,27 +53,44 @@ func Run(ctx context.Context, dir string, args []string) (*internal.TestResults,
...
+			it.out, it.err = runCommand(dir, "mongo", filepath.Join(volume, f))
+			ch <- it
+		}(f)
+	}
+
+	for range files {
```
Comment (AlekSi) [inline internal/jstest/jstest.go:75]:
```suggestion
	for it := range ch {
```

**Review by AlekSi — COMMENTED** [review]
Please use `sync.WaitGroup` to wait for goroutines. And also please run `task lint`

**Comment by b1ron** [comment]
@AlekSi linter is working locally but still getting this [failure](https://github.com/FerretDB/dance/actions/runs/4477690547/jobs/7869535097#step:18:98) on CI. PTAL.

**Comment by AlekSi** [comment]
Please run `task fmt`

File: internal/jstest/jstest.go [hunk 4]
```diff
@@ -53,27 +55,56 @@ func Run(ctx context.Context, dir string, args []string) (*internal.TestResults,
...
 
-		if err == nil {
-			res.TestResults[testName] = internal.TestResult{
-				Status: internal.Pass,
-				Output: string(output),
+	// closer
+	go func() {
+		wg.Wait()
+		close(ch)
+	}()
```
Comment (AlekSi) [inline internal/jstest/jstest.go:87]:
```suggestion
	wg.Wait()

	close(ch)
```

No need to use a goroutine there

File: internal/jstest/jstest.go [hunk 5]
```diff
@@ -53,27 +55,56 @@ func Run(ctx context.Context, dir string, args []string) (*internal.TestResults,
...
+
+	for it := range ch {
+		if it.err != nil {
+			var exitErr *exec.ExitError
+			if !errors.As(it.err, &exitErr) {
+				return nil, it.err
 			}
+		}
 
+		if it.err != nil {
```
Comment (AlekSi) [inline internal/jstest/jstest.go:97]: It is the same check as above

File: internal/jstest/jstest.go (hunk 4 above)
Comment (b1ron) [reply internal/jstest/jstest.go]: Nice!

File: internal/jstest/jstest.go (hunk 5 above)
Comment (b1ron) [reply internal/jstest/jstest.go]: Nice catch!

File: internal/jstest/jstest.go [hunk 6]
```diff
@@ -53,27 +55,52 @@ func Run(ctx context.Context, dir string, args []string) (*internal.TestResults,
...
 	volume := "tests"
-	for _, testName := range files {
-		output, err := runCommand(dir, "mongo", filepath.Join(volume, testName))
-		if err != nil {
-			if _, ok := err.(*exec.ExitError); !ok {
-				return nil, err
+
+	ch := make(chan *item, len(files))
+
+	var wg sync.WaitGroup
```
Comment (AlekSi) [inline internal/jstest/jstest.go:68]: We run all files in parallel there, without an upper bound. That's fine now, but something we could improve in the future

**Review by AlekSi — APPROVED** [review]

**Pull request merged**